flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/AbstractStateChangeLogger.java

```diff
@@ -87,11 +98,27 @@ protected void log(
             ThrowingConsumer<DataOutputViewStreamWrapper, IOException> dataWriter,
             Ns ns)
             throws IOException {
-        // todo: log metadata (FLINK-22808)
+        logMetaIfNeeded();
         stateChangelogWriter.append(
                 keyContext.getCurrentKeyGroupIndex(), serialize(op, ns, dataWriter));
     }
 
+    private void logMetaIfNeeded() throws IOException {
+        if (metaDataWritten) {
+            return;
+        }
+        stateChangelogWriter.append(
+                COMMON_KEY_GROUP,
```

Comment:
A negative keygroup number here is used to make writer write metadata in the very beginning of its current batch. That way, on recovery metadata is be read before the data changes, even if they were made to a smaller keygroup or come in because of rescaling.

I'd add a version of `append`  without `keygroup` to `StateChangelogWriter`, but probably there are some better alternatives?